components/places/android/src/main/java/org/mozilla/places/Bookmarks.kt

```diff
@@ -0,0 +1,407 @@
+     * the one passed in (however, it will be the same according to the
+     * [URL standard](https://url.spec.whatwg.org/)).
+     *
+     * @param url The url to search for.
+     * @return A list of bookmarks that have the requested URL.
+     */
+    fun getBookmarksWithUrl(url: String): List<BookmarkTreeNode>
+}
+
+/**
+ * The methods provided by a bookmarks connection with write capabilities.
+ */
+interface BookmarkWriteInterface : BookmarkReadInterface {
+
+    /**
+     * Delete the bookmark with the provided GUID.
+     *
+     * If the requested bookmark is a folder, all children of
+     * bookmark are deleted as well, recursively.
+     *
+     * @param guid The GUID of the bookmark to delete
+     * @return Whether or not the bookmark existed.
+     */
+    fun deleteBookmark(guid: String): Boolean
+
+    /**
+     * Create a bookmark folder, returning it's guid.
+     *
+     * @param parentGUID The GUID of the (soon to be) parent of this bookmark.
+     * @param title The title of the folder.
+     * @param position The index where to insert the record inside
+     *                 it's parent. If not provided, this item will
+     *                 be appended.
+     * @return The GUID of the newly inserted bookmark folder.
+     */
+    fun createFolder(
+            parentGUID: String,
+            title: String,
+            position: Int? = null
+    ): String
+
+    /**
+     * Create a bookmark separator, returning it's guid.
+     *
+     * @param parentGUID The GUID of the (soon to be) parent of this bookmark.
+     * @param position The index where to insert the record inside
+     *                 it's parent. If not provided, this item will
+     *                 be appended.
+     * @return The GUID of the newly inserted bookmark separator.
+     */
+    fun createSeparator(
+            parentGUID: String,
+            position: Int? = null
+    ): String
+
+    /**
+     * Create a bookmark item, returning it's guid.
+     *
+     * @param parentGUID The GUID of the (soon to be) parent of this bookmark.
+     * @param url The URL to bookmark
+     * @param title The title of the new bookmark, if any.
+     * @param position The index where to insert the record inside
+     *                 it's parent. If not provided, this item will
+     *                 be appended.
+     * @return The GUID of the newly inserted bookmark item.
+     */
+    fun createBookmarkItem(
+            parentGUID: String,
+            url: String,
+            title: String?,
+            position: Int? = null
+    ): String
+
+    /**
+     * Update a bookmark to the provided info.
+     *
+     * @param guid Guid of the bookmark to update
+     * @param info The changes to make to the listed bookmark.
+     */
+    fun updateBookmark(guid: String, info: BookmarkUpdateInfo)
+}
+
+/**
+ * Information describing the changes to make in order to update a bookmark.
+ */
+data class BookmarkUpdateInfo(
```

Comment:
For data integrity purposes, what do you think about also requiring current `parentGUID`, `position`, etc whenever these fields are asked to be modified?

Specifically, I'm thinking about user actions interleaving in time with a running sync; user may ask to move a bookmark (or modify it in some other way) that's also been modified remotely (and those changes have been synced).

This is what some of fennec's APIs look like; current state of the world is specified along with a requested change, and if there's a mismatch (e.g. the world moved on, somehow), operation is cancelled.

Reply:
I've talked with grisha and I'm going to change this API to be a bit different. Essentailly:

- move and update will be different calls
- update will take something type-speciric
- move will allow (optionally) passing in what it expects the current position and parent are, and we'll fail if it doesn't match.

This is all only in the public API, the calls under the hood will remain the same.

Reply:
Ditto re: not doing this, it's more complexity than I think is warranted and I don't see what benefit it would have.